### PR TITLE
Auth: Should redirect to login when anonymous enabled and URL with different org than anonymous specified 

### LINKED ISSUE #26120: A more helpful information, when accessing a dashboard while not logged in
Hey everybody,

we used Grafana without anonymous login. Now we have a organization for visitors in our Grafana instance.
Before this change users got redirected to the login page, if they are not logged in.
If they now want to access a dashboard, where you need to be logged in, you only get the error message "Dashboard not found".
Many users can't find the problem with this message and ask for help.

Is there a chance to add more information to this error message like ""Dashboard not found, maybe you need to login."?

Thanks in advance!

Best regards,
Patrick

### PR DESCRIPTION
**Pull request by marefr**
**What this PR does / why we need it**:
If anonymous access is enabled for an org and there are multiple
orgs. When a non-authenticated user requests a page that requires 
user to be logged in and orgId query string is set in the request URL 
to an org that not equals the anonymous org, the user should be 
redirected to the login page.

**Which issue(s) this PR fixes**:
Fixes #26120 

**Special notes for your reviewer**:
As soon as I realized the problem described in the referenced issue I understood that it's the same problem I've experienced several times on for example play.grafana.org. That's why I took a stab at trying to resolve this.

Tried to convert tests to standard Go test library, but there's just too many changes needed for the `middlewareScenario` so suggest trying to fix that some other time.

### CODE REVIEW

File: pkg/middleware/auth_test.go [hunk 1]
```diff
@@ -32,6 +34,46 @@ func TestMiddlewareAuth(t *testing.T) {
 			})
 		})
 
+		Convey("Anonymous auth enabled", func() {
+			setting.AnonymousEnabled = true
+			setting.AnonymousOrgName = "test"
```
Comment (aknuds1) [inline pkg/middleware/auth_test.go:39]: Global state changes should be cleaned up.
```suggestion
			origEnabled := setting.AnonymousEnabled
			t.Cleanup(func() {
				setting.AnonymousEnabled = origEnabled
			})
			origName := setting.AnonymousOrgName
			t.Cleanup(func() {
				setting.AnonymousOrgName = origName
			})
			setting.AnonymousEnabled = true
			setting.AnonymousOrgName = "test"
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Global state changes in tests should be cleaned up, to avoid potential interference with other tests.

**Review by papagian — CHANGES_REQUESTED** [review]
I have tested the following scenarios and all work perfect! I only have some small suggestions.

I have two organisations (org1 and org2) and I have enabled anonymous access for org1.
- A user navigates to a dashboard in org2 and is redirected to the login page. 
- A user navigates to a dashboard in org1 and is navigated to the dashboard.
- A user navigates to a dashboard in org1 (with `forceLogin=true`) and redirected to the login page.

I have also tested without enabling  anonymous access and the the user is always redirected to the login page.

File: pkg/middleware/auth.go [hunk 2]
```diff
@@ -94,6 +94,14 @@ func Auth(options *AuthOptions) macaron.Handler {
 			if err == nil {
 				forceLogin = forceLoginParam
 			}
+
+			if !forceLogin {
+				orgIDValue := c.Req.URL.Query().Get("orgId")
+				orgID, err := strconv.ParseInt(orgIDValue, 10, 32)
+				if err == nil && orgID > 0 && orgID != c.OrgId {
```
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/middleware/auth.go:101]: I copied it from https://github.com/grafana/grafana/blob/957c88eaca0bfd666ad525ef607df4d60c4c4ea2/pkg/middleware/org_redirect.go#L19-L20

Yes, it's probably safer.

**Review by aknuds1 — APPROVED** [review]
LGTM

File: pkg/middleware/auth.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (marefr) [reply pkg/middleware/auth.go:101]: I changed to int64 in both places

**Review by papagian — APPROVED** [review]

**Comment by aknuds1** [comment]
This needs a merge with master.

**Pull request merged**